How should I test my own package?

Is there any official pattern for testing a 3rd party package on Pyodide, that does not intend to have their package become part of Pyodide?

I'm trying to run my own package, [dit_cli](https://github.com/ditabase/dit-cli), in Pyodide. This was actually quite easy to set up, you can see a very rough demo [here](https://dev.ditabase.io/demo.html). This is a 100% pure Python package, I'm not using the science stack, I just want to run Python/WASM efficiently in the browser.

Now I would like to set up testing, following the Pyodide pattern using Selenium. I've got the basics working, but a lot of what `conftest.py` does is very unclear, and apparently not intended for testing 3rd party packages.

- Should I package Pyodide with my project, just for testing?
- Should I just copy `conftest.py` and the rest of the testing system? What else do I need to know if I do that?

Ah, okay. This took me a bit to follow, but you're saying I should drop Selenium, package the tests with the `.whl`, and then just run the tests in a real browser? This is definitely easier in some ways.

I've done a version of this on my demo page:
```JavaScript
async function startPyodide() {
  await loadPyodide({
    indexURL: "https://cdn.jsdelivr.net/pyodide/v0.17.0/full/",
    fullStdLib: false,
  });
  await pyodide.runPythonAsync(`
    import pytest
    import micropip
    await micropip.install('setuptools')
    await micropip.install('${window.location.origin}/js/dit_cli-0.5.1-py3-none-any.whl')
    pytest.main(['--pyargs', 'dit_cli'])
  `);
}
```
However, I get an error with pytest, it seems like some things are not imported. I tried running the tests for `matplotlib` and got a very similar error message.
<details>
  <summary>Error log</summary>
  
  ```
============================= test session starts ==============================
platform emscripten -- Python 3.8.2, pytest-3.6.3, py-1.9.0, pluggy-0.13.1
rootdir: /, inifile:

collecting 0 items / 1 errors                                                  
collected 0 items / 1 errors                                                   

==================================== ERRORS ====================================
______________________________ ERROR collecting  _______________________________
lib/python3.8/site-packages/_pytest/config/__init__.py:372: in _getconftestmodules
    return self._path2confmods[path]
E   KeyError: local('/lib/python3.8/site-packages/dit_cli/tests')

During handling of the above exception, another exception occurred:
lib/python3.8/site-packages/_pytest/config/__init__.py:403: in _importconftest
    return self._conftestpath2mod[conftestpath]
E   KeyError: local('/lib/python3.8/site-packages/dit_cli/tests/conftest.py')

During handling of the above exception, another exception occurred:
lib/python3.8/site-packages/_pytest/config/__init__.py:409: in _importconftest
    mod = conftestpath.pyimport()
lib/python3.8/site-packages/py/_path/local.py:704: in pyimport
    __import__(modname)
lib/python3.8/site-packages/_pytest/assertion/rewrite.py:226: in load_module
    py.builtin.exec_(co, mod.__dict__)
lib/python3.8/site-packages/dit_cli/tests/conftest.py:3: in <module>
    from _pytest.reports import TestReport
E   ModuleNotFoundError: No module named '_pytest.reports'

During handling of the above exception, another exception occurred:
lib/python3.8/site-packages/py/_path/common.py:383: in visit
    for x in Visitor(fil, rec, ignore, bf, sort).gen(self):
lib/python3.8/site-packages/py/_path/common.py:424: in gen
    dirs = self.optsort([p for p in entries
lib/python3.8/site-packages/py/_path/common.py:425: in <listcomp>
    if p.check(dir=1) and (rec is None or rec(p))])
lib/python3.8/site-packages/_pytest/main.py:511: in _recurse
    ihook = self.gethookproxy(path)
lib/python3.8/site-packages/_pytest/main.py:413: in gethookproxy
    my_conftestmodules = pm._getconftestmodules(fspath)
lib/python3.8/site-packages/_pytest/config/__init__.py:386: in _getconftestmodules
    mod = self._importconftest(conftestpath)
lib/python3.8/site-packages/_pytest/config/__init__.py:417: in _importconftest
    raise ConftestImportFailure(conftestpath, sys.exc_info())
E   _pytest.config.ConftestImportFailure: ModuleNotFoundError("No module named '_pytest.reports'")
E     File "/lib/python3.8/site-packages/_pytest/assertion/rewrite.py", line 226, in load_module
E       py.builtin.exec_(co, mod.__dict__)
E     File "/lib/python3.8/site-packages/dit_cli/tests/conftest.py", line 3, in <module>
E       from _pytest.reports import TestReport
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 1.04 seconds ============================
  ```
  
</details>

--- a/docs/project/changelog.md
+++ b/docs/project/changelog.md
@@ -69,6 +69,13 @@ substitutions:
 
 - {{Fix}} pillow now correctly encodes/decodes JPEG image format. {pr}`1818`
 
+### Micellaneous
+
+- {{Fix}} Patched emscripten to make the system calls to duplicate file
+  descriptors closer to posix-compliant. In particular, this fixes the use of
+  `dup` on pipes and temporary files, as needed by `pytest`.
+  {pr}`1823`
+
 ## Version 0.18.0
 
 _August 3rd, 2021_

--- a/emsdk/patches/fix-dup.patch
+++ b/emsdk/patches/fix-dup.patch
@@ -70,13 +70,6 @@ index 96d2ec0c3..0001624ec 100644
    },
    __sys_pipe2__nothrow: true,
    __sys_pipe2__proxy: false,
-diff --git a/tests/third_party/posixtestsuite b/tests/third_party/posixtestsuite
-index 2e3f88374..a5e7ff5f9 160000
---- a/tests/third_party/posixtestsuite
-+++ b/tests/third_party/posixtestsuite
-@@ -1 +1 @@
--Subproject commit 2e3f883749a5344618b02b09dd19a6dc9e4b70a2
-+Subproject commit a5e7ff5f906c9a8f18d70f6472c4055bc311c34d
 -- 
 2.17.1
 

--- a/emsdk/patches/fix-dup.patch
+++ b/emsdk/patches/fix-dup.patch
@@ -19,8 +19,8 @@ This patch only partially resolves the problems with `dup` (it doesn't fully dup
 
 diff --git a/emsdk/upstream/emscripten/src/library_syscall.js b/emsdk/upstream/emscripten/src/library_syscall.js
 index 96d2ec0c3..0001624ec 100644
---- a/src/library_syscall.js
-+++ b/src/library_syscall.js
+--- a/emsdk/upstream/emscripten/src/library_syscall.js
++++ b/emsdk/upstream/emscripten/src/library_syscall.js
 @@ -137,10 +137,10 @@ var SyscallsLibrary = {
        }
        return 0;

--- a/emsdk/patches/fix-dup.patch
+++ b/emsdk/patches/fix-dup.patch
@@ -1,0 +1,82 @@
+From 52c0b163bad96cde2081c734680b9ec485fea908 Mon Sep 17 00:00:00 2001
+From: Hood <hood@mit.edu>
+Date: Wed, 8 Sep 2021 17:49:15 -0700
+Subject: [PATCH] Fix dup
+
+---
+ src/library_syscall.js           | 12 ++++++------
+ tests/third_party/posixtestsuite |  2 +-
+ 2 files changed, 7 insertions(+), 7 deletions(-)
+
+This should fix two problems with the `dup` system calls:
+1. Pipes cannot be duplicated (https://github.com/emscripten-core/emscripten/issues/14640)
+2. `TemporaryFiles` cannot be duplicated (https://github.com/emscripten-core/emscripten/issues/15012)
+
+Both of these issues cause trouble with pytest. There is an upstream pull request that would fix this problem:
+https://github.com/emscripten-core/emscripten/pull/9396/files
+
+This patch only partially resolves the problems with `dup` (it doesn't fully duplicate the changes in the emscripten PR) but I think it will be good enough to fix pytest.
+
+diff --git a/emsdk/upstream/emscripten/src/library_syscall.js b/emsdk/upstream/emscripten/src/library_syscall.js
+index 96d2ec0c3..0001624ec 100644
+--- a/src/library_syscall.js
++++ b/src/library_syscall.js
+@@ -137,10 +137,10 @@ var SyscallsLibrary = {
+       }
+       return 0;
+     },
+-    doDup: function(path, flags, suggestFD) {
++    doDup: function(stream, suggestFD) {
+       var suggest = FS.getStream(suggestFD);
+       if (suggest) FS.close(suggest);
+-      return FS.open(path, flags, 0, suggestFD, suggestFD).fd;
++      return FS.createStream(stream, suggestFD, suggestFD).fd;
+     },
+     doReadv: function(stream, iov, iovcnt, offset) {
+       var ret = 0;
+@@ -380,7 +380,7 @@ var SyscallsLibrary = {
+   },
+   __sys_dup: function(fd) {
+     var old = SYSCALLS.getStreamFromFD(fd);
+-    return FS.open(old.path, old.flags, 0).fd;
++    return FS.createStream(old, 0).fd;
+   },
+   __sys_pipe__deps: ['$PIPEFS'],
+   __sys_pipe: function(fdPtr) {
+@@ -472,7 +472,7 @@ var SyscallsLibrary = {
+   __sys_dup2: function(oldfd, suggestFD) {
+     var old = SYSCALLS.getStreamFromFD(oldfd);
+     if (old.fd === suggestFD) return suggestFD;
+-    return SYSCALLS.doDup(old.path, old.flags, suggestFD);
++    return SYSCALLS.doDup(old, suggestFD);
+   },
+   __sys_getppid__nothrow: true,
+   __sys_getppid__proxy: false,
+@@ -1167,7 +1167,7 @@ var SyscallsLibrary = {
+           return -{{{ cDefine('EINVAL') }}};
+         }
+         var newStream;
+-        newStream = FS.open(stream.path, stream.flags, 0, arg);
++        newStream = FS.createStream(stream, arg);
+         return newStream.fd;
+       }
+       case {{{ cDefine('F_GETFD') }}}:
+@@ -1403,7 +1403,7 @@ var SyscallsLibrary = {
+     assert(!flags);
+ #endif
+     if (old.fd === suggestFD) return -{{{ cDefine('EINVAL') }}};
+-    return SYSCALLS.doDup(old.path, old.flags, suggestFD);
++    return SYSCALLS.doDup(old, suggestFD);
+   },
+   __sys_pipe2__nothrow: true,
+   __sys_pipe2__proxy: false,
+diff --git a/tests/third_party/posixtestsuite b/tests/third_party/posixtestsuite
+index 2e3f88374..a5e7ff5f9 160000
+--- a/tests/third_party/posixtestsuite
++++ b/tests/third_party/posixtestsuite
+@@ -1 +1 @@
+-Subproject commit 2e3f883749a5344618b02b09dd19a6dc9e4b70a2
++Subproject commit a5e7ff5f906c9a8f18d70f6472c4055bc311c34d
+-- 
+2.17.1
+

--- a/src/tests/test_pyodide.py
+++ b/src/tests/test_pyodide.py
@@ -164,9 +164,18 @@ def test_dup_pipe():
     # See https://github.com/emscripten-core/emscripten/issues/14640
     import os
 
-    [fdr, fdw] = os.pipe()
-    os.dup(fdr)
-    os.dup(fdw)
+    [fdr1, fdw1] = os.pipe()
+    fdr2 = os.dup(fdr1)
+    fdw2 = os.dup2(fdw1, 50)
+    # Closing any of fdr, fdr2, fdw, or fdw2 will currently destroy the pipe.
+    # This bug is fixed upstream:
+    # https://github.com/emscripten-core/emscripten/pull/14685
+    s1 = b"some stuff"
+    s2 = b"other stuff to write"
+    os.write(fdw1, s1)
+    assert os.read(fdr2, 100) == s1
+    os.write(fdw2, s2)
+    assert os.read(fdr1, 100) == s2
 
 
 @run_in_pyodide
@@ -176,8 +185,13 @@ def test_dup_temp_file():
     from tempfile import TemporaryFile
 
     tf = TemporaryFile()
-    os.dup(tf.fileno())
-    os.dup(tf.fileno())
+    fd1 = os.dup(tf.fileno())
+    fd2 = os.dup2(tf.fileno(), 50)
+    s = b"hello there!"
+    tf.write(s)
+    # This next assertion actually demonstrates a bug in dup: the correct value
+    # to return should be b"".
+    assert os.read(fd1, 10) == s
 
 
 @pytest.mark.skip_pyproxy_check

--- a/src/tests/test_pyodide.py
+++ b/src/tests/test_pyodide.py
@@ -2,6 +2,7 @@ import pytest
 from pathlib import Path
 import sys
 from textwrap import dedent
+from pyodide_build.testing import run_in_pyodide
 
 sys.path.append(str(Path(__file__).resolve().parents[2] / "src" / "py"))
 
@@ -156,6 +157,27 @@ def test_eval_code_locals():
     with pytest.raises(NameError):
         eval_code("invalidate_caches()", globals, globals)
     eval_code("invalidate_caches()", globals, locals)
+
+
+@run_in_pyodide
+def test_dup_pipe():
+    # See https://github.com/emscripten-core/emscripten/issues/14640
+    import os
+
+    [fdr, fdw] = os.pipe()
+    os.dup(fdr)
+    os.dup(fdw)
+
+
+@run_in_pyodide
+def test_dup_temp_file():
+    # See https://github.com/emscripten-core/emscripten/issues/15012
+    import os
+    from tempfile import TemporaryFile
+
+    tf = TemporaryFile()
+    os.dup(tf.fileno())
+    os.dup(tf.fileno())
 
 
 @pytest.mark.skip_pyproxy_check

--- a/src/tests/test_pyodide.py
+++ b/src/tests/test_pyodide.py
@@ -184,14 +184,14 @@ def test_dup_temp_file():
     import os
     from tempfile import TemporaryFile
 
-    tf = TemporaryFile()
+    tf = TemporaryFile(buffering=0)
     fd1 = os.dup(tf.fileno())
     fd2 = os.dup2(tf.fileno(), 50)
     s = b"hello there!"
     tf.write(s)
     # This next assertion actually demonstrates a bug in dup: the correct value
     # to return should be b"".
-    assert os.read(fd1, 10) == s
+    assert os.read(fd1, 50) == s
 
 
 @pytest.mark.skip_pyproxy_check

